### PR TITLE
[image_picker_web] Listens to file input cancel event.

### PR DESCRIPTION
**Pull request by ditman**
## Changes

This PR listens to the `cancel` event from the `input type=file` used by the web implementation of the image_picker plugin, so apps don't end up endlessly awaiting for a file that will never come **in modern browsers** (Chrome 113, Safari 16.4, or newer). _Same API as https://github.com/flutter/packages/pull/3683._

Additionally, this PR:

* Removes all code and tests mentioning `PickedFile`. (Deprecated years ago, and unused since https://github.com/flutter/packages/pull/4285) **(Breaking change)**
* Updates README to mention `XFile` which is the current return type of the package.
* Updates the dependency on `image_picker_platform_interface` to `^2.9.0`.
  * Implements all non-deprecated methods from the interface, and makes deprecated methods use the fresh ones.
  * Updates tests.

### Issues

* Fixes https://github.com/flutter/flutter/issues/92176

### Testing

* Added integration testing coverage for the 'cancel' event.
* Tested manually in Chrome with the example app running on web.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/image_picker/image_picker_for_web/CHANGELOG.md
Comment (ditman) [inline packages/image_picker/image_picker_for_web/CHANGELOG.md:6]: This has been only around for a few weeks, but I don't want to postpone the cleanup much longer. Should I make this PR a major version bump, so it's guaranteed that nobody ends up in a state of having an out-of-date core plugin, but a "too new" web impl?)

Comment (stuartmorgan) [reply packages/image_picker/image_picker_for_web/CHANGELOG.md:6]: Sorry, I thought I'd responded to this already. Removing deprecated code is a breaking change, so requires a major version bump, otherwise you will actively break people with older versions of the app-facing package.

> but I wouldn't want to postpone this cleanup too much.

We haven't done any of the cleanup in the lower layers; we should do it bottom-up for the non-user-facing parts.

> Or can I assume that people migrated to the new API years ago, with the deprecation?

You can't assume anything that the resolver will allow to happen won't happen :)

File: packages/image_picker/image_picker_for_web/README.md
Comment (stuartmorgan) [inline packages/image_picker/image_picker_for_web/README.md:9]: (These README changes are all fine; we don't need to still README the deprecated implementation APIs.)

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/image_picker/image_picker_for_web/README.md
Comment (ditman) [reply packages/image_picker/image_picker_for_web/README.md:9]: > we don't need to still README the deprecated implementation APIs.

Having trouble parse that bit? Please rephrase?

**Comment by ditman** [comment]
I've updated the README to mention the current methods (not deprecated ones), and marked this change as a major version to prevent any issues with package resolution (I'll send a PR updating `image_picker` to v.3 of the web implementation after this lands). PTAL @stuartmorgan!

File: packages/image_picker/image_picker_for_web/CHANGELOG.md
Comment (stuartmorgan) [reply packages/image_picker/image_picker_for_web/CHANGELOG.md]: I checked, and we haven't even formally deprecated the old methods in the platform interface package, which makes removing the implementations here worrying since we aren't doing anything to ensure we don't call it anywhere.

I think we should deprecate the methods first in a prequel PR, and then update the min version of the platform interface this package requires to the one that deprecates them. I'll make that PR now, then we can update this one after that lands.

Comment (ditman) [reply packages/image_picker/image_picker_for_web/CHANGELOG.md]: @stuartmorgan the methods may not be deprecated on the `platform_interface`, but everything that returned a `PickedFile` was marked as deprecated ~2 years ago in the core plugin https://github.com/flutter/plugins/pull/4073, and eventually removed here https://github.com/flutter/packages/pull/4285, right?

AFAIK, nothing that returns a `PickedFile` is being called anymore by the plugin as of version `^1.0` (could it? yes)

(This deprecation started on the front-end, of the package, I guess :P)

As for landing this in sequence with the deprecation in the platform_interface: no problem either! :)

Comment (stuartmorgan) [reply packages/image_picker/image_picker_for_web/CHANGELOG.md]: > @stuartmorgan the methods may not be deprecated on the `platform_interface`, but everything that returned a `PickedFile` was marked as deprecated ~2 years ago in the core plugin [flutter/plugins#4073](https://github.com/flutter/plugins/pull/4073), and eventually removed here #4285, right?
> 
> AFAIK, nothing that returns a `PickedFile` is being called anymore by the plugin as of version `^1.0` (could it? yes)

Correct, it's not that we are calling this in the app-facing package's implementation, it's that there's nothing structural that would prevent us from calling it in the future (and as I found when I tried to do the deprecation, nothing preventing other locations in our codebase from calling it).

> (This deprecation started on the front-end, of the package, I guess :P)

There are two distinct API layers: the client-facing API, and the platform interface API. Deprecating things in the client-facing API is unrelated in general to deprecating things in the platform interface (although they happen to be more closely linked than usual here).

My concern is that the platform implementations should be implementing the platform interface API as it is expressed formally, rather than tailoring the implementations to how current clients are calling it.

File: packages/image_picker/image_picker_for_web/pubspec.yaml
Comment (stuartmorgan) [inline packages/image_picker/image_picker_for_web/pubspec.yaml:6]: Let's update the `image_picker_platform_interface` to `^2.9.0` so that this can't be used with a version that hasn't deprecated the methods you removed.

File: packages/image_picker/image_picker_for_web/README.md
Comment (stuartmorgan) [inline packages/image_picker/image_picker_for_web/README.md:72]: Nit: Should that be `Blob`?

**Review by stuartmorgan — APPROVED** [review]
LGTM modulo the version constraint.

File: packages/image_picker/image_picker_for_web/README.md
Comment (ditman) [reply packages/image_picker/image_picker_for_web/README.md:72]: Yeah, MDN uses [both](https://developer.mozilla.org/en-US/docs/Web/API/Blob), but the non-monospace version looks like exceptions to the rule.

**Comment by ditman** [comment]
Hey @stuartmorgan, here's the refresh to the `image_picker_web` to implement the latest platform interface! PTAL (and change/merge at will)

**Review by stuartmorgan — APPROVED** [review]
StillLGTM

**Pull request merged**
